From 3887a75fba2c07a25aadce39f179f6679d7f8958 Mon Sep 17 00:00:00 2001
From: Patrick Ohly <patrick.ohly@intel.com>
Date: Fri, 8 Apr 2016 11:20:18 +0200
Subject: [PATCH] pam_motd.c: fall back to system's default motd

In a stateless distro, /etc is empty and the system defaults for
aspects typically configured via /etc are under
/usr/share/defaults/etc instead.

Support that also with pam_motd by first checking /etc and if not
found, falling back to the system defaults.

Signed-off-by: Patrick Ohly <patrick.ohly@intel.com>
---
 modules/pam_motd/pam_motd.c | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/modules/pam_motd/pam_motd.c b/modules/pam_motd/pam_motd.c
index ff9b169..0d2fa9f 100644
--- a/modules/pam_motd/pam_motd.c
+++ b/modules/pam_motd/pam_motd.c
@@ -21,6 +21,7 @@
 #include <sys/stat.h>
 #include <pwd.h>
 #include <syslog.h>
+#include <errno.h>
 
 #include <security/_pam_macros.h>
 #include <security/pam_ext.h>
@@ -80,7 +81,8 @@ int pam_sm_open_session(pam_handle_t *pamh, int flags,
     if (motd_path == NULL)
 	motd_path = default_motd;
 
-    while ((fd = open(motd_path, O_RDONLY, 0)) >= 0) {
+    while ((fd = open(motd_path, O_RDONLY, 0)) >= 0 ||
+           (motd_path == default_motd && errno == ENOENT && ((fd = open("/usr/share/defaults/etc/motd", O_RDONLY, 0)) >= 0))) {
 	struct stat st;
 
 	/* fill in message buffer with contents of motd */
-- 
2.1.4

